From a3de65d5d1861f755ced7cad291fbbd4f1b8ef51 Mon Sep 17 00:00:00 2001
From: Owen W. Taylor <otaylor@fishsoup.net>
Date: Sat, 22 Aug 2009 15:00:57 -0400
Subject: [PATCH] Should set RestartStyleHint to RestartIfRunning when replaced

This reverts most of commit abbd057eb967e6ab462ffe305f41b2b04d417b25;

 - It's fine to call meta_session_shutdown() after the display
   is closed, since it's talking over the ICE connection
 - We should not call warn_about_lame_clients_and_finish_interact()
   unless we are interacting with the window manager in a session
   save.

However, the part of abbd057 that fixed accessing freed memory was
fixing a real problem; this patches does the same thing in a simpler
way by fixing an obvious type in meta_display_close() where it was
NULL'ing out the local variable 'display' rather than the global
variable 'the_display' and adding keeping the check in meta_finalize()
that was added in abbd057.

The order of calling meta_session_shutdown() and
calling meta_display_close() is reverted back to the old order to
make it clear that it's OK if the display way already closed previously.

http://bugzilla.gnome.org/show_bug.cgi?id=588119
diff -Nur -x '*.orig' -x '*~' metacity-2.27.1/src/core/display.c metacity-2.27.1.new/src/core/display.c
--- metacity-2.27.1/src/core/display.c	2009-09-10 16:47:16.000000000 +1000
+++ metacity-2.27.1.new/src/core/display.c	2009-09-10 16:47:17.000000000 +1000
@@ -926,7 +926,7 @@
     meta_compositor_destroy (display->compositor);
   
   g_free (display);
-  display = NULL;
+  the_display = NULL;
 
   meta_quit (META_EXIT_SUCCESS);
 }
@@ -4763,13 +4763,10 @@
       meta_verbose ("Got selection clear for screen %d on display %s\n",
                     screen->number, display->name);
       
-      meta_display_unmanage_screen (&display, 
+      meta_display_unmanage_screen (display, 
                                     screen,
                                     event->xselectionclear.time);
 
-      if (!display)
-        the_display = NULL;
-
       /* display and screen may both be invalid memory... */
       
       return;
@@ -4791,12 +4788,10 @@
 }
 
 void
-meta_display_unmanage_screen (MetaDisplay **displayp,
+meta_display_unmanage_screen (MetaDisplay *display,
                               MetaScreen  *screen,
                               guint32      timestamp)
 {
-  MetaDisplay *display = *displayp;
-
   meta_verbose ("Unmanaging screen %d on display %s\n",
                 screen->number, display->name);
   
@@ -4806,10 +4801,7 @@
   display->screens = g_slist_remove (display->screens, screen);
 
   if (display->screens == NULL)
-    {
-      meta_display_close (display, timestamp);
-      *displayp = NULL;
-    }
+    meta_display_close (display, timestamp);
 }
 
 void
diff -Nur -x '*.orig' -x '*~' metacity-2.27.1/src/core/display-private.h metacity-2.27.1.new/src/core/display-private.h
--- metacity-2.27.1/src/core/display-private.h	2009-09-09 06:55:35.000000000 +1000
+++ metacity-2.27.1.new/src/core/display-private.h	2009-09-10 16:47:17.000000000 +1000
@@ -329,7 +329,7 @@
 void          meta_display_grab                (MetaDisplay *display);
 void          meta_display_ungrab              (MetaDisplay *display);
 
-void          meta_display_unmanage_screen     (MetaDisplay **display,
+void          meta_display_unmanage_screen     (MetaDisplay *display,
                                                 MetaScreen  *screen,
                                                 guint32      timestamp);
 
diff -Nur -x '*.orig' -x '*~' metacity-2.27.1/src/core/main.c metacity-2.27.1.new/src/core/main.c
--- metacity-2.27.1/src/core/main.c	2009-09-09 06:55:35.000000000 +1000
+++ metacity-2.27.1.new/src/core/main.c	2009-09-10 16:47:17.000000000 +1000
@@ -361,12 +361,11 @@
 meta_finalize (void)
 {
   MetaDisplay *display = meta_get_display();
-
-  meta_session_shutdown ();
-
   if (display)
     meta_display_close (display,
                         CurrentTime); /* I doubt correct timestamps matter here */
+
+  meta_session_shutdown ();
 }
 
 static void
diff -Nur -x '*.orig' -x '*~' metacity-2.27.1/src/core/session.c metacity-2.27.1.new/src/core/session.c
--- metacity-2.27.1/src/core/session.c	2009-09-09 06:55:35.000000000 +1000
+++ metacity-2.27.1.new/src/core/session.c	2009-09-10 16:47:17.000000000 +1000
@@ -376,14 +376,6 @@
   SmProp *props[1];
   char hint = SmRestartIfRunning;
 
-  if (!meta_get_display ())
-    {
-      meta_verbose ("Cannot close session because there is no display");
-      return;
-    }
-
-  warn_about_lame_clients_and_finish_interact (FALSE);
-
   if (session_connection == NULL)
     return;
   
